flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalTemporalJoin.scala

```diff
@@ -0,0 +1,206 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.nodes.physical.stream
+
+import org.apache.flink.table.api.ValidationException
+import org.apache.flink.table.planner.calcite.FlinkTypeFactory
+import org.apache.flink.table.planner.plan.nodes.common.CommonPhysicalJoin
+import org.apache.flink.table.planner.plan.nodes.exec.{ExecEdge, ExecNode}
+import org.apache.flink.table.planner.plan.nodes.exec.stream.StreamExecTemporalJoin
+import org.apache.flink.table.planner.plan.nodes.exec.utils.JoinSpec
+import org.apache.flink.table.planner.plan.utils.TemporalJoinUtil.{TEMPORAL_JOIN_CONDITION, TEMPORAL_JOIN_CONDITION_PRIMARY_KEY}
+import org.apache.flink.table.planner.plan.utils.TemporalJoinUtil
+import org.apache.flink.util.Preconditions.checkState
+
+import org.apache.calcite.plan.{RelOptCluster, RelTraitSet}
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.core.{Join, JoinRelType}
+import org.apache.calcite.rex._
+
+import java.util.Optional
+
+import scala.collection.JavaConverters._
+
+/**
+ * Stream physical node for temporal table join (FOR SYSTEM_TIME AS OF) and
+ * temporal TableFunction join (LATERAL TemporalTableFunction(proctime)).
+ *
+ * <p>The legacy temporal table function join is the subset of temporal table join,
+ * the only difference is the validation, we reuse most same logic here.
+ */
+class StreamPhysicalTemporalJoin(
+    cluster: RelOptCluster,
+    traitSet: RelTraitSet,
+    leftRel: RelNode,
+    rightRel: RelNode,
+    condition: RexNode,
+    joinType: JoinRelType)
+  extends CommonPhysicalJoin(cluster, traitSet, leftRel, rightRel, condition, joinType)
+  with StreamPhysicalRel {
+
+  override def requireWatermark: Boolean = {
+    TemporalJoinUtil.isRowTimeJoin(joinSpec)
+  }
+
+  override def copy(
+      traitSet: RelTraitSet,
+      conditionExpr: RexNode,
+      left: RelNode,
+      right: RelNode,
+      joinType: JoinRelType,
+      semiJoinDone: Boolean): Join = {
+    new StreamPhysicalTemporalJoin(
+      cluster,
+      traitSet,
+      left,
+      right,
+      conditionExpr,
+      joinType)
+  }
+
+  override def translateToExecNode(): ExecNode[_] = {
+    val textualRepresentation = this.toString
+    val rexBuilder = cluster.getRexBuilder
+    val isTemporalFunctionJoin =
+        TemporalJoinUtil.isTemporalFunctionJoin(rexBuilder, joinInfo)
+
+    val leftFieldCount = getLeft.getRowType.getFieldCount
+    val temporalJoinConditionExtractor = new TemporalJoinConditionExtractor(
+      textualRepresentation,
+      leftFieldCount,
+      joinSpec,
+      cluster.getRexBuilder,
+      isTemporalFunctionJoin)
+    val remainingNonEquiJoinPredicates =
+      temporalJoinConditionExtractor.apply(joinSpec.getNonEquiCondition.orElse(null)) match {
+        case condition if condition != null => Optional.of(condition)
+        case _ => Optional.empty().asInstanceOf[Optional[RexNode]]
+      }
```

Comment:
nit: these line can be simplified as `Optional.ofNullable(
      temporalJoinConditionExtractor.apply(joinSpec.getNonEquiCondition.orElse(null)))`